examples/exex/kona/src/main.rs

```diff
@@ -0,0 +1,244 @@
+    std::env::var(var).unwrap_or_else(|_| panic!("{var} must be set")).parse().unwrap()
+}
+
+fn info_from_header(block: &reth::primitives::SealedBlock) -> BlockInfo {
+    BlockInfo {
+        hash: block.hash(),
+        number: block.number,
+        timestamp: block.timestamp,
+        parent_hash: block.parent_hash,
+    }
+}
+
+type LocalPipeline =
+    DerivationPipeline<LocalAttributesQueue<LocalDataProvider>, AlloyL2ChainProvider>;
+
+type LocalDataProvider = EthereumDataSource<providers::LocalChainProvider, blobs::ExExBlobProvider>;
+
+type LocalAttributesBuilder =
+    StatefulAttributesBuilder<providers::LocalChainProvider, AlloyL2ChainProvider>;
+
+type LocalAttributesQueue<DAP> = AttributesQueue<
+    BatchQueue<
+        ChannelReader<
+            ChannelBank<FrameQueue<L1Retrieval<DAP, L1Traversal<providers::LocalChainProvider>>>>,
+        >,
+        AlloyL2ChainProvider,
+    >,
+    LocalAttributesBuilder,
+>;
+
+pub async fn follow_safe_head(mut ctx: ExExContext<impl FullNodeComponents>) -> eyre::Result<()> {
+    let cfg = Arc::new(OP_MAINNET_CONFIG);
+    let mut chain_provider = providers::LocalChainProvider::new();
+    chain_provider.insert_block_info(BlockInfo {
+        hash: cfg.genesis.l1.hash,
+        number: cfg.genesis.l1.number,
+        timestamp: Default::default(),
+        parent_hash: Default::default(),
+    });
+    let mut l2_provider = AlloyL2ChainProvider::new_http(new_req_url(L2_RPC_URL), cfg.clone());
+    let attributes =
+        StatefulAttributesBuilder::new(cfg.clone(), l2_provider.clone(), chain_provider.clone());
+    let beacon_client = OnlineBeaconClient::new_http(BEACON_URL.to_string());
+    let blob_provider =
+        OnlineBlobProvider::<_, SimpleSlotDerivation>::new(beacon_client, None, None);
+    let blob_store = Arc::new(Mutex::new(blobs::InMemoryBlobProvider::new()));
+    let blob_provider = blobs::ExExBlobProvider::new(Arc::clone(&blob_store), blob_provider);
+    let dap = EthereumDataSource::new(chain_provider.clone(), blob_provider, &cfg);
+    let mut cursor = l2_provider
+        .l2_block_info_by_number(cfg.genesis.l2.number)
+        .await
+        .expect("Failed to fetch genesis L2 block info for pipeline cursor");
+    let tip = chain_provider
+        .block_info_by_number(cursor.l1_origin.number)
+        .await
+        .expect("Failed to fetch genesis L1 block info for pipeline tip");
+    let validator = validation::OnlineValidator::new_http(
+        new_req_url(L2_RPC_URL),
+        cfg.canyon_time.unwrap_or_default(),
+    );
+    let mut pipeline: LocalPipeline = PipelineBuilder::new()
+        .rollup_config(cfg.clone())
+        .dap_source(dap)
+        .l2_chain_provider(l2_provider.clone())
+        .chain_provider(chain_provider.clone())
+        .builder(attributes)
+        .origin(tip)
+        .build();
+    let mut derived_attributes_count = 0;
+    let mut cursor_map: HashMap<BlockInfo, L2BlockInfo> = HashMap::new();
```

Comment:
IMO this should be a `struct` with an `async` method on it, and all the initializations should happen in `new`